### PR TITLE
Implementation of kernelSR

### PR DESCRIPTION
**Pull request by riccardo-rende**
This PR implements the kernel formulation of Stochastic Reconfiguration.

We provide a new driver, `netket.experimental.VMC_SRt,`. It reproduces **exactly** the dynamics of the standard `netket.driver.VMC` with `nk.optimizer.SR(solver=netket.optimizer.solver.solvers.solve))`, but with the inversion of N_samples x N_samples matrix instead of a N_params x N_params one. We add a test (test/driver/test_srt.py) to show that this is the case.

N.B. It requires `rearrange` from `einops.`

### CODE REVIEW

File: netket/experimental/driver/vmc_kernel.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        """
+        Initializes the driver class.
+
+        Args:
+            hamiltonian: The Hamiltonian of the system.
+            optimizer: Determines how optimization steps are performed given the
+                    bare energy gradient.
+            diag_shift: The diagonal shift of the stochastic reconfiguration matrix.
+                    Typical values are 1e-4 ÷ 1e-3.
+        """
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:114]:
```
            jacobian_mode: differentiation mode, must be one of `real` or `complex`. `real` truncates
                  the imaginary part of the wavefunction, while `complex` does not. If left unspecified the
                  correct one is selected.

        """
```

File: netket/experimental/driver/vmc_kernel.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    ):
+        """
+        Initializes the driver class.
+
+        Args:
+            hamiltonian: The Hamiltonian of the system.
+            optimizer: Determines how optimization steps are performed given the
+                    bare energy gradient.
+            diag_shift: The diagonal shift of the stochastic reconfiguration matrix.
+                    Typical values are 1e-4 ÷ 1e-3.
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:113]:
```suggestion
            diag_shift: The diagonal shift of the stochastic reconfiguration matrix.
                    Typical values are 1e-4 ÷ 1e-3. Can also be an optax schedule.
```

File: netket/experimental/driver/vmc_kernel.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+import jax.numpy as jnp
+import jax.scipy as jsp
+from netket.driver.vmc import VMC
+
+import netket.jax as nkjax
+from netket.operator import AbstractOperator
+
+from netket.utils.mpi import rank, n_nodes
+from netket.utils.mpi import mpi_gather_jax, mpi_alltoall_jax, mpi_reduce_sum_jax, mpi_scatter_jax, mpi_allreduce_sum_jax
+
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:15]:
```suggestion
from netket.utils.types import ScalarOrSchedule

```

File: netket/experimental/driver/vmc_kernel.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    :math:`\delta \\theta = \\tau X(X^TX + \lambda \mathbb{I}_{2M})^{-1} f`,
+
+    where :math:`X \in R^{P \\times 2M}` is the concatenation of the real and imaginary part 
+    of the centered Jacobian, with P the number of parameters and M the number of samples.
+    The vector f is the concatenation of the real and imaginary part of the centered local
+    energy. Note that, to compute the updates, it is sufficient to invert an :math:`M\\times M` matrix
+    instead of a :math:`P\\times P` one. As a consequence, this formulation is useful
+    in the typical deep learning regime where :math:`P \gg M`.
+    
+    See R.Rende, L.L.Viteritti, L.Bardone, F.Becca and S.Goldt (https://arxiv.org/abs/2310.05715)
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:88]: this should be an hyperlink (the syntax is terrible... I know),

```suggestion
    See `R.Rende, L.L.Viteritti, L.Bardone, F.Becca and S.Goldt <https://arxiv.org/abs/2310.05715>`_
```

File: netket/experimental/driver/vmc_kernel.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+class VMC_kernelSR(VMC):
+    """
+    Energy minimization using Variational Monte Carlo (VMC) and the kernel 
+    formulation of Stochastic Reconfiguration (SR). This approach lead to
+    *exactly* the same parameter updates of the standard SR with a
+    diagonal shift regularization. For this reason, it is equivalent to the standard 
+    nk.driver.VMC with the preconditioner nk.optimizer.SR(solver=netket.optimizer.solver.solvers.solve))
+    In the kernel SR framework, the updates of the parameters can be written as:
+
+    :math:`\delta \\theta = \\tau X(X^TX + \lambda \mathbb{I}_{2M})^{-1} f`,
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:79]: this makes it big, not inline
```suggestion
    .. math::
        \delta \theta = \tau X(X^TX + \lambda \mathbb{I}_{2M})^{-1} f,
```

File: netket/experimental/driver/vmc_kernel.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    updates = O_L.T @ aus_vector
+    updates, token = mpi_allreduce_sum_jax(updates, token=token)
+    
+    return -updates 
+
+inv_default_solver = lambda A, b: jnp.linalg.inv(A) @ b
+linear_solver = lambda A, b: jsp.linalg.solve(A, b, assume_a="pos")
+
+class VMC_kernelSR(VMC):
+    """
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:71]: this avoids all those double dashes (please remove some if I missed them)
```suggestion
    r"""
```

File: netket/experimental/driver/vmc_kernel.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    Energy minimization using Variational Monte Carlo (VMC) and the kernel 
+    formulation of Stochastic Reconfiguration (SR). This approach lead to
+    *exactly* the same parameter updates of the standard SR with a
+    diagonal shift regularization. For this reason, it is equivalent to the standard 
+    nk.driver.VMC with the preconditioner nk.optimizer.SR(solver=netket.optimizer.solver.solvers.solve))
+    In the kernel SR framework, the updates of the parameters can be written as:
+
+    :math:`\delta \\theta = \\tau X(X^TX + \lambda \mathbb{I}_{2M})^{-1} f`,
+
+    where :math:`X \in R^{P \\times 2M}` is the concatenation of the real and imaginary part
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:81]:
```suggestion
    where :math:`X \in R^{P \times 2M}` is the concatenation of the real and imaginary part 
```

File: netket/experimental/driver/vmc_kernel.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    nk.driver.VMC with the preconditioner nk.optimizer.SR(solver=netket.optimizer.solver.solvers.solve))
+    In the kernel SR framework, the updates of the parameters can be written as:
+
+    :math:`\delta \\theta = \\tau X(X^TX + \lambda \mathbb{I}_{2M})^{-1} f`,
+
+    where :math:`X \in R^{P \\times 2M}` is the concatenation of the real and imaginary part 
+    of the centered Jacobian, with P the number of parameters and M the number of samples.
+    The vector f is the concatenation of the real and imaginary part of the centered local
+    energy. Note that, to compute the updates, it is sufficient to invert an :math:`M\\times M` matrix
+    instead of a :math:`P\\times P` one. As a consequence, this formulation is useful
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:85]:
```suggestion
    instead of a :math:`P\times P` one. As a consequence, this formulation is useful
```

Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:85]:
```suggestion
    instead of a :math:`P\times P` one. As a consequence, this formulation is useful
```

File: netket/experimental/driver/vmc_kernel.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    def __init__(
+        self,
+        hamiltonian: AbstractOperator,
+        optimizer,
+        diag_shift,
+        *args,
+        linear_solver_fn: Callable[[jax.Array, jax.Array], jax.Array] = linear_solver,
+        jacobian_mode = None,
+        variational_state = None,
+        **kwargs,
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:103]: Let's remove optional arguments. Also add documentation to types
```suggestion
        self,
        hamiltonian: AbstractOperator,
        optimizer,
        *,
        diag_shift: ScalarOrSchedule,
        linear_solver_fn: Callable[[jax.Array, jax.Array], jax.Array] = linear_solver,
        jacobian_mode: Optional[str] = None,
        variational_state: VariationalState = None
        )
```

File: netket/experimental/driver/vmc_kernel.py (hunk 2 above)
Comment (PhilipVinc) [reply netket/experimental/driver/vmc_kernel.py:113]: Please also add a documentation about what signature should `linear_solver_fn` have

File: netket/experimental/driver/vmc_kernel.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+                    Typical values are 1e-4 ÷ 1e-3.
+        """
+        super().__init__(hamiltonian, optimizer, variational_state=variational_state)
+
+        self._ham = hamiltonian.collect()  # type: AbstractOperator
+        _, self.unravel_params_fn = ravel_pytree(self.state.parameters)
+        self.diag_shift = diag_shift
+        self.jacobian_mode = jacobian_mode
+        self.linear_solver_fn = linear_solver_fn
+
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:122]: This makes an automatic default, and better errors..
```suggestion
    @property
    def jacobian_mode(self) -> str:
        """
        The mode used to compute the jacobian of the variational state. Can be `'real'`
        or `'complex'`.

        Real mode truncates imaginary part of the wavefunction, while `complex` does not.
        This internally uses :func:`netket.jax.jacobian`. See that function for a more
        complete documentation.
        """
        return self._jacobian_mode

    @jacobian_mode.setter
    def jacobian_mode(self, mode: Optional[str]):
        if mode is None:
            mode = nkjax.jacobian_default_mode(
                self.state._apply_fun, 
                self.state.parameters, 
                self.state.model_state,
                self.state.samples,
                warn=False
                )

        if mode not in ["complex", "real"]:
            raise ValueError("`jacobian_mode` only supports 'real' for real-valued wavefunctions and"
                "'complex'.\n\n"
                "`holomorphic` is not yet supported, but could be contributed in the future."
                )
        self._jacobian_mode = mode

```

File: netket/experimental/driver/vmc_kernel.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        self._loss_stats = e_mean
+
+        de = jnp.conj(local_energy - e_mean)
+
+        if self.jacobian_mode is None:
+            mode = "complex"
+        else:
+            #* mode='complex' is the most general; mode='holomorphic' could be implemented
+            assert self.jacobian_mode in ["complex", "real"], "Jacobian mode must be 'complex' or 'real'"
+            mode = self.jacobian_mode
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:146]: let's handle this above in the constructor.
```suggestion
```

File: netket/experimental/driver/vmc_kernel.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+        jacobians = nkjax.jacobian(self.state._apply_fun,
+                                self.state.parameters,
+                                self.samples.squeeze(), 
+                                self.state.model_state,
+                                mode=mode,
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:152]:
```suggestion
                                mode=self.jacobian_mode,
```

File: netket/experimental/driver/vmc_kernel.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+                                dense=True,
+                                center=True) #* jaxcobians is centered
+
+        updates = kernel_SR(jacobians, de, self.diag_shift, mode, self.linear_solver_fn)
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:156]: let's make this internal for now...
```suggestion
        updates = kernel_SR(jacobians, de, self.diag_shift, self.jacobian_mode, self._linear_solver_fn)
```

File: netket/experimental/driver/vmc_kernel.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+            diag_shift: The diagonal shift of the stochastic reconfiguration matrix.
+                    Typical values are 1e-4 ÷ 1e-3.
+        """
+        super().__init__(hamiltonian, optimizer, variational_state=variational_state)
+
+        self._ham = hamiltonian.collect()  # type: AbstractOperator
+        _, self.unravel_params_fn = ravel_pytree(self.state.parameters)
+        self.diag_shift = diag_shift
+        self.jacobian_mode = jacobian_mode
+        self.linear_solver_fn = linear_solver_fn
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:121]: let's mark this internal for now. 
```suggestion
        self._linear_solver_fn = linear_solver_fn
```

File: netket/experimental/driver/vmc_kernel.py [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+
+        updates = kernel_SR(jacobians, de, self.diag_shift, mode, self.linear_solver_fn)
+
+        self._dp = self.unravel_params_fn(updates)
+
+        return self._dp
+    
+    @property
+    def samples(self):
+        return self.state._samples #! is there a better way?
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:164]: why do you need this? I think you can avoid it.
```suggestion
    
```

File: netket/experimental/driver/vmc_kernel.py [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+        if self.jacobian_mode is None:
+            mode = "complex"
+        else:
+            #* mode='complex' is the most general; mode='holomorphic' could be implemented
+            assert self.jacobian_mode in ["complex", "real"], "Jacobian mode must be 'complex' or 'real'"
+            mode = self.jacobian_mode
+
+        jacobians = nkjax.jacobian(self.state._apply_fun,
+                                self.state.parameters,
+                                self.samples.squeeze(),
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_kernel.py:150]: can't you just do that?
```suggestion
                                self.state.samples.squeeze(), 
```

**Review by PhilipVinc — COMMENTED** [review]
Seems good! Thanks. I cleaned up a bit the interface and removed the einops dependency.

There might be a few broken things but this should still run.
If you commit my suggestions, than I can move this to the experimental module and merge.

File: netket/experimental/driver/vmc_srt.py [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+        Args:
+            n_steps (int): Number of steps to perform.
+        """
+
+        self.state.reset()
+
+        # Compute the local energy estimator and average Energy
+        local_energy = self.state.local_estimators(self._ham).squeeze()
+
+        e_mean = local_energy.mean()
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_srt.py:181]: @riccardo-rende Are you sure here it's a mean, and not a mean over MPI as well?

Comment (PhilipVinc) [reply netket/experimental/driver/vmc_srt.py]: I think it was wrong. I changed it to an mpi average. it is passing tests under MPI now.

Comment (PhilipVinc) [reply netket/experimental/driver/vmc_srt.py]: please confirm

File: test/driver/test_srt.py [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+import optax
+
+import pytest
+
+from netket.experimental.driver import VMC_SRt
+from netket.optimizer.solver.solvers import solve
+from netket.utils import mpi
+
+
+class RBM(nn.Module):
```
Comment (gcarleo) [inline test/driver/test_srt.py:30]: Do we have tests to check that SR.t gives the same updates as SR ?

**Review by PhilipVinc — APPROVED** [review]
This looks good to me now.
Tests are also passing over MPI so I believe the implementation to be correct, and I added a few extra ones checking that the parameters obtained from SR and SRt are indeed identical. 
Thank you!

If @riccardo-rende confirms that the missing mpi operation was just a typo, I think we can merge.

File: test/driver/test_srt.py [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+        seed=0,
+        sampler_seed=0,
+    )
+
+    return H, opt, vstate
+
+
+def test_SRt_vs_linear_solver():
+    """
+    nk.driver.VMC_kernelSR must give **exactly** the same dynamics as nk.driver.VMC with nk.optimizer.SR
```
Comment (gcarleo) [inline test/driver/test_srt.py:98]: ah ok sorry I didn't see this...

**Review by gcarleo — APPROVED** [review]

File: netket/experimental/driver/vmc_srt.py (hunk 17 above)
Comment (riccardo-rende) [reply netket/experimental/driver/vmc_srt.py]: Yes, it was wrong! I was also able to test the new version using MPI and it works! Thank you very much.

**Comment by gcarleo** [comment]
supersedes, and closes #1432

**Comment by chrisrothUT** [comment]
Thanks [riccardo-rende](https://github.com/riccardo-rende) for implementing this. This implementation gives one of the benefits of MinSR (avoiding the need for regularization) but it still instantiates the Jacobian in memory so I'm not sure it's a complete solution for large models. 

Along the lines of these new SR papers, isn't it possible to just solve for the gradient update directly by solving this linear equation?

(J - <J>) dθ = E-<E>

I bring this up because SR is quadratic in parameters, MinSR is quadratic in samples, but this appears to be linear in both.
